controller/lib/hal/hal_stm32.cpp

```diff
@@ -0,0 +1,1008 @@
+         break;
+   }
+
+   ADC_Regs *adc = reinterpret_cast<ADC_Regs *>(ADC_BASE);
+   adc->adc[0].seq[0] = channel << 6;
+
+   // Clear the EOC flag
+   adc->adc[0].stat = 4;
+
+   // Start the conversion
+   adc->adc[0].ctrl |= 4;
+
+   // Wait for the end of the conversion
+   while( !(adc->adc[0].stat & 4) ){}
+
+   // Return the result
+   return adc->adc[0].data;
+}
+
+/******************************************************************
+ * PWM outputs
+ *
+ * The following four outputs could be driven
+ * as PWM outputs:
+ *
+ * PA8  - Timer 1 Channel 1 - heater control
+ * PA11 - Timer 1 Channel 4 - solenoid
+ * PB3  - Timer 2 Channel 2 - blower control
+ * PB4  - Timer 3 Channel 1 - buzzer
+ *
+ * For now I'll just set up the blower since that's the only
+ * one called out in the HAL
+ *
+ * These timers are documented in chapters 26 and 27 of the reference
+ * manual.
+ *****************************************************************/
+static void InitPwmOut()
+{
+   // The PWM frequency isn't mentioned anywhere that I can find, so
+   // I'm just picking a reasonable number.  This can be refined later
+   const int pwmFreqHz  = 20000;
+
+   EnableClock( TIMER2_BASE );
+
+   // Connect PB3 to timer 2
+   GPIO_PinAltFunc( GPIO_B_BASE, 3, 1 );
+
+   TimerRegs *tmr = reinterpret_cast<TimerRegs *>(TIMER2_BASE);
+
+   // Set the frequency
+   tmr->reload = (CPU_FREQ / pwmFreqHz)-1;
+
+   // Configure channel 2 in PWM output mode
+   tmr->ccMode[0] = 0x6800;
+
+   tmr->ccEnable = 0x10;
+
+   // Start with 0% duty cycle
+   tmr->compare[1] = 0;
+
+   // Load the shadow registers
+   tmr->event = 1;
+
+   // Start the counter
+   tmr->ctrl[0] = 0x81;
+}
+
+// Set the PWM period.  Currently the value is 0 to 255 which is
+// a shame.  We should change this to a float, say with a value
+// of 0.0 to 1.0 to make better use of our resolution
+void HalApi::analogWrite(PwmPin pin, int value) {
+
+   // Convert the value to a float
+   float duty =  value * (1.0/255);
+
+   TimerRegs *tmr;
+   int chan;
+   switch( pin )
+   {
+      case PwmPin::BLOWER:
+         tmr = reinterpret_cast<TimerRegs *>(TIMER2_BASE);
+         chan = 1;
+         break;
+   }
+
+   tmr->compare[chan] = tmr->reload * duty;
+}
+
+/******************************************************************
+ * Serial port to GUI
+ * Chapter 38 of the reference manual defines the USART registers.
+ *****************************************************************/
+
+// This class is a generic circular buffer for byte sized data.
+// It could probably go in it's own header outside the HAL, we
+// would just need to add interrupt suspend/restore to the HAL
+// definition.
+//
+// Note that this class is used from both the main line of code
+// and the interrupt handlers, so it needs to be thread safe.
+// I'm disabling interrupts during the critical sections to
+// ensure that's the case.
+template <int N> class CircBuff
+{
+   volatile uint8_t buff[N];
+   volatile int head, tail;
```

Comment:
I think the volatile keyword is important here.  It's basically telling the compiler that head and tail can magically change on their own, so it shouldn't optimize out it's access to them.  Marking the buffer as volatile is probably not necessary, it's pretty unlikely that the compiler could gain anything from optimizing access to the buffer data anyway.

I'll look into additional adornments to the assembly functions though.  Can't be too carefully, especially as compilers keep getting smarter.

Reply:
> I think the volatile keyword is important here. It's basically telling the compiler that head and tail can magically change on their own, so it shouldn't optimize out it's access to them. 

The thing about volatile is, it doesn't force an ordering of those reads/writes with respect to *anything else* in the system.  That is, other reads/writes can still be reordered across volatile reads/writes, and in particular, *two volatile operations (to different addresses) can be reordered with respect to each other*.

In the circular buffer, we're using interrupts as a mutex.  You want to treat the "disable interrupts" function as though it writes arbitrary memory (so that right after interrupts are disabled, you re-read everything), and you want to treat "enable interrupts" as though it reads arbitrary memory (so that right before you enable interrupts you flush all outstanding writes).

Volatile on the variables doesn't necessarily get you this; it's relying on the compiler not reordering the volatile reads/writes with the "asm volatile", and that's not written down anywhere, I have no idea what it does.  Marking the asm volatile as a memory barrier gets you this for sure.  But then you don't need volatile on the variables.

Up to you if you want to keep it.  All I care really is that we don't get a false sense of security from it.